pandas/core/apply.py

```diff
@@ -1281,8 +1288,59 @@ def apply(self):
     def transform(self):
         raise NotImplementedError
 
+    def agg_or_apply_list_like(
+        self, op_name: Literal["agg", "apply"]
+    ) -> DataFrame | Series:
+        obj = self.obj
+        kwargs = self.kwargs
+        if op_name == "apply":
+            kwargs = {**kwargs, "by_row": False}
+
+        if getattr(obj, "axis", 0) == 1:
+            raise NotImplementedError("axis other than 0 is not supported")
+
+        if obj._selected_obj.ndim == 1:
+            # For SeriesGroupBy this matches _obj_with_exclusions
+            selected_obj = obj._selected_obj
+        else:
+            selected_obj = obj._obj_with_exclusions
+
+        with com.temp_setattr(obj, "as_index", True):
+            keys, results = self.compute_list_like(op_name, selected_obj, kwargs)
+        result = self.wrap_results_list_like(keys, results)
+        return result
+
+    def agg_or_apply_dict_like(
+        self, op_name: Literal["agg", "apply"]
+    ) -> DataFrame | Series:
+        assert op_name in ["agg", "apply"]
+
+        obj = self.obj
+        kwargs = {}
+        if op_name == "apply":
+            by_row = "_compat" if self.by_row else False
+            kwargs.update({"by_row": by_row})
+
+        if getattr(obj, "axis", 0) == 1:
+            raise NotImplementedError("axis other than 0 is not supported")
+
+        selected_obj = obj._selected_obj
+        selection = obj._selection
+
+        # Numba Groupby engine/engine-kwargs passthrough
+        engine = self.kwargs.get("engine", None)
+        engine_kwargs = self.kwargs.get("engine_kwargs", None)
+        kwargs.update({"engine": engine, "engine_kwargs": engine_kwargs})
+
+        with com.temp_setattr(obj, "as_index", True):
+            result_index, result_data = self.compute_dict_like(
+                op_name, selected_obj, selection, kwargs
+            )
+        result = self.wrap_results_dict_like(selected_obj, result_index, result_data)
+        return result
+
 
-class ResamplerWindowApply(Apply):
+class ResamplerWindowApply(GroupByApply):
```

Comment:
> and rename this to `ResamplerWindowApply`?

It is named `ResamplerWindowApply`, so I'm not sure what you're asking.

The main reason to separate this off now is that GroupBy requires the context (to use `as_index=True`) whereas this does not. 

In general, I'm thinking of removing the class structure in this file, keeping a single class (something like `ApplyData` or maybe `UdfData`) that stores the state we need to pass around and making everything else module-level methods. Currently, the use of inheritance makes it hard to understand the code (both for humans and IDEs).

The state (`ApplyData` or `UdfData`) would have `obj`, `selected_obj` (what I would rename to `data`), `func`, `args`, `kwargs`, and any other state derived from these that we reuse.

What do you think of moving toward this structure?

Reply:
> It is named `ResamplerWindowApply`, so I'm not sure what you're asking.

Sorry, I meant just `WindowApply`, but you answered my question (i.e. the difference is `as_index=True`), thanks.

These new methods are very similar to the methods on `GroupbyApply`, the difference being only the numba kwargs and the context manager. Can you look again if it's possible to avoid the repetition?

> What do you think of moving toward this structure?

Fro a surface view if looks to me like it would be simplest to have the current `Apply` class only be used by `Series` and `DataFrame` and let the groupby/window cases be handled in a separate structure.  That would simplify the type hints a lot, for example, and maybe the code also. Is that the plan you're describing, or do you have something else in mind?

Reply:
> Can you look again if it's possible to avoid the repetition?

An alternative is to introduce a `condition` argument to `temp_setattr` so that optionally setting an attribute is a bit easier. I've taken this approach in the latest revision.

> Is that the plan you're describing, or do you have something else in mind?

I was more thinking of still having groupby / window / resample in `core.apply`, just not having the inheritance structure. I attempted this today and it's a bit of a mixed bag. The code may be more straightforward to follow, but there is also more branching. I think your improvements that will take hold in 3.0 and some general cleanups will improve this code alone; I'm going to hold off on pursuing this more for now.